[FR] Proper EEPROM support on Fysetc S6

### Description

Add support for the onboard EEPROM for Fysetc S6

### Feature Workflow

According to the documentation for Fysetc S6 there is an onboard 24LC16 EEPROM
When compiling with PRINTCOUNTER enabled, an error is issued:
```
Marlin\src\HAL\STM32\../../inc/../HAL/STM32/inc/SanityCheck.h:44:4: warning: #warning "FLASH_EEPROM_EMULATION may cause long delays when writing and should not be used while printing." [-Wcpp]
   44 |   #warning "FLASH_EEPROM_EMULATION may cause long delays when writing and should not be used while printing."
      |    ^~~~~~~
Marlin\src\HAL\STM32\../../inc/../HAL/STM32/inc/SanityCheck.h:45:4: error: #error "Disable PRINTCOUNTER or choose another EEPROM emulation."

This is a bit wierd, as the controller has on-board EEPROM.

Also, a warning is issued on:
Marlin\src\HAL\STM32\../../inc/../HAL/STM32/inc/SanityCheck.h:44:4: warning: #warning "FLASH_EEPROM_EMULATION may cause long delays when writing and should not be used while printing." [-Wcpp]
   44 |   #warning "FLASH_EEPROM_EMULATION may cause long delays when writing and should not be used while printing."
```
But no FLASH_EEPROM_EMULATION definitions exist on configuration.h or configuration_adv.h

#### Additional Information

Schematics for Fysetc S6 can be found on https://github.com/FYSETC/FYSETC-S6

Tested with a pull of the current bugfix on June 26th, 2020

@sjasonsmith - do you have any input on whether #define IIC_BL24CXX_EEPROM should be used or not ?
I had it enabled in a compile attempt, but get this error:

Marlin\src\lcd/dwin/eeprom_BL24CXX.h:29:10: fatal error: libmaple/gpio.h: No such file or directory

But if the define shouldn't be used, there is n o need to dig further into it...

Not sure whether to open a new ticket. I just tried with the bugfix branch from today (20200727), and with changing pins_FYSETC_S6.h to have:

```
//
// EEPROM Emulation
//
#if NO_EEPROM_SELECTED
  //#define FLASH_EEPROM_EMULATION
  //#define SRAM_EEPROM_EMULATION
  #define I2C_EEPROM
#endif

#if ENABLED(FLASH_EEPROM_EMULATION)
  // Decrease delays and flash wear by spreading writes across the
  // 128 kB sector allocated for EEPROM emulation.
  #define FLASH_EEPROM_LEVELING
#elif ENABLED(I2C_EEPROM)
  #define IIC_EEPROM_SDA PB9
  #define IIC_EEPROM_SCL PB8
  #define MARLIN_EEPROM_SIZE 0x800               // 2KB
#endif
```

Compile of the bugfix branch from 20200726 worked OK.

bugfix-2.0.x from 20200727 gives me (with same modification to pins file and same configuration.h/configuration_adv.h files):

```
Compiling .pio\build\FYSETC_S6\src\src\HAL\STM32\eeprom_impl.cpp.o
Marlin\src\HAL\STM32\eeprom_impl.cpp: In static member function 'static bool PersistentStore::write_data(int&, const uint8_t*, size_t, uint16_t*)':
Marlin\src\HAL\STM32\eeprom_impl.cpp:48:33: error: invalid conversion from 'uint8_t*' {aka 'unsigned char*'} to 'uint32_t' {aka 'long unsigned int'} [-fpermissive]
   48 |       if (v != eeprom_read_byte(p)) {
      |                                 ^
      |                                 |
      |                                 uint8_t* {aka unsigned char*}
In file included from C:\users\brian.ipsen\.platformio\packages\framework-arduinoststm32@3.10700.191028\cores\arduino/board.h:18,
                 from C:\users\brian.ipsen\.platformio\packages\framework-arduinoststm32@3.10700.191028\cores\arduino/wiring.h:41,
                 from C:\users\brian.ipsen\.platformio\packages\framework-arduinoststm32@3.10700.191028\cores\arduino/Arduino.h:32,
                 from Marlin\src\HAL\STM32\../../inc/../HAL/./STM32/../shared/Marduino.h:36,
                 from Marlin\src\HAL\STM32\../../inc/../HAL/./STM32/HAL.h:28,
                 from Marlin\src\HAL\STM32\../../inc/../HAL/HAL.h:26,
                 from Marlin\src\HAL\STM32\../../inc/MarlinConfig.h:30,
                 from Marlin\src\HAL\STM32\eeprom_impl.cpp:25:
C:\users\brian.ipsen\.platformio\packages\framework-arduinoststm32@3.10700.191028\cores\arduino\stm32/stm32_eeprom.h:63:41: note:   initializing argument 1 of 'uint8_t eeprom_read_byte(uint32_t)'
   63 | uint8_t eeprom_read_byte(const uint32_t pos);
      |                          ~~~~~~~~~~~~~~~^~~
Marlin\src\HAL\STM32\eeprom_impl.cpp:49:27: error: invalid conversion from 'uint8_t*' {aka 'unsigned char*'} to 'uint32_t' {aka 'long unsigned int'} [-fpermissive]
   49 |         eeprom_write_byte(p, v);
      |                           ^
      |                           |
      |                           uint8_t* {aka unsigned char*}
In file included from C:\users\brian.ipsen\.platformio\packages\framework-arduinoststm32@3.10700.191028\cores\arduino/board.h:18,
                 from C:\users\brian.ipsen\.platformio\packages\framework-arduinoststm32@3.10700.191028\cores\arduino/wiring.h:41,
                 from C:\users\brian.ipsen\.platformio\packages\framework-arduinoststm32@3.10700.191028\cores\arduino/Arduino.h:32,
                 from Marlin\src\HAL\STM32\../../inc/../HAL/./STM32/../shared/Marduino.h:36,
                 from Marlin\src\HAL\STM32\../../inc/../HAL/./STM32/HAL.h:28,
                 from Marlin\src\HAL\STM32\../../inc/../HAL/HAL.h:26,
                 from Marlin\src\HAL\STM32\../../inc/MarlinConfig.h:30,
                 from Marlin\src\HAL\STM32\eeprom_impl.cpp:25:
C:\users\brian.ipsen\.platformio\packages\framework-arduinoststm32@3.10700.191028\cores\arduino\stm32/stm32_eeprom.h:64:33: note:   initializing argument 1 of 'void eeprom_write_byte(uint32_t, uint8_t)'
   64 | void eeprom_write_byte(uint32_t pos, uint8_t value);
      |                        ~~~~~~~~~^~~
Marlin\src\HAL\STM32\eeprom_impl.cpp:50:30: error: invalid conversion from 'uint8_t*' {aka 'unsigned char*'} to 'uint32_t' {aka 'long unsigned int'} [-fpermissive]
   50 |         if (eeprom_read_byte(p) != v) {
      |                              ^
      |                              |
      |                              uint8_t* {aka unsigned char*}
In file included from C:\users\brian.ipsen\.platformio\packages\framework-arduinoststm32@3.10700.191028\cores\arduino/board.h:18,
                 from C:\users\brian.ipsen\.platformio\packages\framework-arduinoststm32@3.10700.191028\cores\arduino/wiring.h:41,
                 from C:\users\brian.ipsen\.platformio\packages\framework-arduinoststm32@3.10700.191028\cores\arduino/Arduino.h:32,
                 from Marlin\src\HAL\STM32\../../inc/../HAL/./STM32/../shared/Marduino.h:36,
                 from Marlin\src\HAL\STM32\../../inc/../HAL/./STM32/HAL.h:28,
                 from Marlin\src\HAL\STM32\../../inc/../HAL/HAL.h:26,
                 from Marlin\src\HAL\STM32\../../inc/MarlinConfig.h:30,
                 from Marlin\src\HAL\STM32\eeprom_impl.cpp:25:
C:\users\brian.ipsen\.platformio\packages\framework-arduinoststm32@3.10700.191028\cores\arduino\stm32/stm32_eeprom.h:63:41: note:   initializing argument 1 of 'uint8_t eeprom_read_byte(uint32_t)'
   63 | uint8_t eeprom_read_byte(const uint32_t pos);
      |                          ~~~~~~~~~~~~~~~^~~
Marlin\src\HAL\STM32\eeprom_impl.cpp: In static member function 'static bool PersistentStore::read_data(int&, uint8_t*, size_t, uint16_t*, bool)':
Marlin\src\HAL\STM32\eeprom_impl.cpp:72:26: error: invalid conversion from 'uint8_t*' {aka 'unsigned char*'} to 'uint32_t' {aka 'long unsigned int'} [-fpermissive]
   72 |         eeprom_read_byte((uint8_t*)pos)
      |                          ^~~~~~~~~~~~~
      |                          |
      |                          uint8_t* {aka unsigned char*}
In file included from C:\users\brian.ipsen\.platformio\packages\framework-arduinoststm32@3.10700.191028\cores\arduino/board.h:18,
                 from C:\users\brian.ipsen\.platformio\packages\framework-arduinoststm32@3.10700.191028\cores\arduino/wiring.h:41,
                 from C:\users\brian.ipsen\.platformio\packages\framework-arduinoststm32@3.10700.191028\cores\arduino/Arduino.h:32,
                 from Marlin\src\HAL\STM32\../../inc/../HAL/./STM32/../shared/Marduino.h:36,
                 from Marlin\src\HAL\STM32\../../inc/../HAL/./STM32/HAL.h:28,
                 from Marlin\src\HAL\STM32\../../inc/../HAL/HAL.h:26,
                 from Marlin\src\HAL\STM32\../../inc/MarlinConfig.h:30,
                 from Marlin\src\HAL\STM32\eeprom_impl.cpp:25:
C:\users\brian.ipsen\.platformio\packages\framework-arduinoststm32@3.10700.191028\cores\arduino\stm32/stm32_eeprom.h:63:41: note:   initializing argument 1 of 'uint8_t eeprom_read_byte(uint32_t)'
   63 | uint8_t eeprom_read_byte(const uint32_t pos);
      |                          ~~~~~~~~~~~~~~~^~~
*** [.pio\build\FYSETC_S6\src\src\HAL\STM32\eeprom_impl.cpp.o] Error 1

```

--- a/Marlin/src/pins/pins.h
+++ b/Marlin/src/pins/pins.h
@@ -644,12 +644,12 @@
   #include "stm32f4/pins_VAKE403D.h"            // STM32F4
 #elif MB(FYSETC_S6)
   #include "stm32f4/pins_FYSETC_S6.h"           // STM32F4                                env:FYSETC_S6
+#elif MB(FYSETC_S6_V2_0)
+  #include "stm32f4/pins_FYSETC_S6_V2_0.h"      // STM32F4                                env:FYSETC_S6
 #elif MB(FLYF407ZG)
   #include "stm32f4/pins_FLYF407ZG.h"           // STM32F4                                env:FLYF407ZG
 #elif MB(MKS_ROBIN2)
   #include "stm32f4/pins_MKS_ROBIN2.h"          // STM32F4                                env:MKS_ROBIN2
-#elif MB(FYSETC_S6_V2_0)
-  #include "stm32f4/pins_FYSETC_S6_V2_0.h"      // STM32F4                                env:FYSETC_S6
 
 //
 // ARM Cortex M7

--- a/Marlin/src/pins/stm32f4/pins_FYSETC_S6.h
+++ b/Marlin/src/pins/stm32f4/pins_FYSETC_S6.h
@@ -41,9 +41,9 @@
 // EEPROM Emulation
 //
 #if NO_EEPROM_SELECTED
-  #define FLASH_EEPROM_EMULATION
+  //#define FLASH_EEPROM_EMULATION
   //#define SRAM_EEPROM_EMULATION
-  //#define I2C_EEPROM
+  #define I2C_EEPROM
 #endif
 
 #if ENABLED(FLASH_EEPROM_EMULATION)
@@ -51,7 +51,7 @@
   // 128 kB sector allocated for EEPROM emulation.
   #define FLASH_EEPROM_LEVELING
 #elif ENABLED(I2C_EEPROM)
-  #define MARLIN_EEPROM_SIZE              0x1000  // 4KB
+  #define MARLIN_EEPROM_SIZE              0x0800  // 2KB
 #endif
 
 //
@@ -62,12 +62,24 @@
 //
 // Limit Switches
 //
-#define X_MIN_PIN                           PB14
-#define X_MAX_PIN                           PA1
-#define Y_MIN_PIN                           PB13
-#define Y_MAX_PIN                           PA2
-#define Z_MIN_PIN                           PA0
-#define Z_MAX_PIN                           PA3
+#ifndef X_MIN_PIN
+  #define X_MIN_PIN                         PB14
+#endif
+#ifndef X_MAX_PIN
+  #define X_MAX_PIN                         PA1
+#endif
+#ifndef Y_MIN_PIN
+  #define Y_MIN_PIN                         PB13
+#endif
+#ifndef Y_MAX_PIN
+  #define Y_MAX_PIN                         PA2
+#endif
+#ifndef Z_MIN_PIN
+  #define Z_MIN_PIN                         PA0
+#endif
+#ifndef Z_MAX_PIN
+  #define Z_MAX_PIN                         PA3
+#endif
 
 //
 // Filament Sensor
@@ -169,14 +181,28 @@
 //
 // Heaters / Fans
 //
-#define HEATER_0_PIN                        PB3
-#define HEATER_1_PIN                        PB4
-#define HEATER_2_PIN                        PB15
-#define HEATER_BED_PIN                      PC8
+#ifndef HEATER_0_PIN
+  #define HEATER_0_PIN                      PB3
+#endif
+#ifndef HEATER_1_PIN
+  #define HEATER_1_PIN                      PB4
+#endif
+#ifndef HEATER_2_PIN
+  #define HEATER_2_PIN                      PB15
+#endif
+#ifndef HEATER_BED_PIN
+  #define HEATER_BED_PIN                    PC8
+#endif
 
-#define FAN_PIN                             PB0
-#define FAN1_PIN                            PB1
-#define FAN2_PIN                            PB2
+#ifndef FAN_PIN
+  #define FAN_PIN                           PB0
+#endif
+#ifndef FAN1_PIN
+  #define FAN1_PIN                          PB1
+#endif
+#ifndef FAN2_PIN
+  #define FAN2_PIN                          PB2
+#endif
 
 //
 // SPI

--- a/Marlin/src/pins/stm32f4/pins_FYSETC_S6.h
+++ b/Marlin/src/pins/stm32f4/pins_FYSETC_S6.h
@@ -42,7 +42,6 @@
 //
 #if NO_EEPROM_SELECTED
   #define FLASH_EEPROM_EMULATION
-  //#define SRAM_EEPROM_EMULATION
   //#define I2C_EEPROM
 #endif
 

--- a/Marlin/src/pins/stm32f4/pins_FYSETC_S6.h
+++ b/Marlin/src/pins/stm32f4/pins_FYSETC_S6.h
@@ -41,9 +41,9 @@
 // EEPROM Emulation
 //
 #if NO_EEPROM_SELECTED
-  //#define FLASH_EEPROM_EMULATION
+  #define FLASH_EEPROM_EMULATION
   //#define SRAM_EEPROM_EMULATION
-  #define I2C_EEPROM
+  //#define I2C_EEPROM
 #endif
 
 #if ENABLED(FLASH_EEPROM_EMULATION)
@@ -62,24 +62,12 @@
 //
 // Limit Switches
 //
-#ifndef X_MIN_PIN
-  #define X_MIN_PIN                         PB14
-#endif
-#ifndef X_MAX_PIN
-  #define X_MAX_PIN                         PA1
-#endif
-#ifndef Y_MIN_PIN
-  #define Y_MIN_PIN                         PB13
-#endif
-#ifndef Y_MAX_PIN
-  #define Y_MAX_PIN                         PA2
-#endif
-#ifndef Z_MIN_PIN
-  #define Z_MIN_PIN                         PA0
-#endif
-#ifndef Z_MAX_PIN
-  #define Z_MAX_PIN                         PA3
-#endif
+#define X_MIN_PIN                           PB14
+#define X_MAX_PIN                           PA1
+#define Y_MIN_PIN                           PB13
+#define Y_MAX_PIN                           PA2
+#define Z_MIN_PIN                           PA0
+#define Z_MAX_PIN                           PA3
 
 //
 // Filament Sensor
@@ -181,28 +169,14 @@
 //
 // Heaters / Fans
 //
-#ifndef HEATER_0_PIN
-  #define HEATER_0_PIN                      PB3
-#endif
-#ifndef HEATER_1_PIN
-  #define HEATER_1_PIN                      PB4
-#endif
-#ifndef HEATER_2_PIN
-  #define HEATER_2_PIN                      PB15
-#endif
-#ifndef HEATER_BED_PIN
-  #define HEATER_BED_PIN                    PC8
-#endif
+#define HEATER_0_PIN                        PB3
+#define HEATER_1_PIN                        PB4
+#define HEATER_2_PIN                        PB15
+#define HEATER_BED_PIN                      PC8
 
-#ifndef FAN_PIN
-  #define FAN_PIN                           PB0
-#endif
-#ifndef FAN1_PIN
-  #define FAN1_PIN                          PB1
-#endif
-#ifndef FAN2_PIN
-  #define FAN2_PIN                          PB2
-#endif
+#define FAN_PIN                             PB0
+#define FAN1_PIN                            PB1
+#define FAN2_PIN                            PB2
 
 //
 // SPI

--- a/Marlin/src/pins/stm32f4/pins_FYSETC_S6_V2_0.h
+++ b/Marlin/src/pins/stm32f4/pins_FYSETC_S6_V2_0.h
@@ -29,7 +29,6 @@
 #if NO_EEPROM_SELECTED
   #undef NO_EEPROM_SELECTED
   //#define FLASH_EEPROM_EMULATION
-  //#define SRAM_EEPROM_EMULATION
   #define I2C_EEPROM
 #endif
 

